### PR TITLE
[Fusilli] Added lit and sample tests for grouped strided ConvWGrad

### PR DESCRIPTION
**Pull request by a-sidorova**
The remaining tests from https://github.com/nod-ai/shark-ai/pull/2599

### CODE REVIEW

File: sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+// LINALG-CHECK:        linalg.yield %[[SUM]] : f32
+// LINALG-CHECK:      } -> tensor<256x128x1x1xf32>
+// LINALG-CHECK:      %[[CAST1:.+]] = tensor.cast %[[RED]] : tensor<256x128x1x1xf32> to tensor<?x?x?x?xf32>
+// LINALG-CHECK:      %[[CAST2:.+]] = tensor.cast %[[CAST1]] : tensor<?x?x?x?xf32> to tensor<256x16x1x1xf32>
+// LINALG-CHECK:      %[[OUT_BUF4:.+]] = tensor.empty() : tensor<256x1x1x16xf32>
+// LINALG-CHECK:      %[[OUT_T4:.+]] = linalg.transpose ins(%[[CAST2]] : tensor<256x16x1x1xf32>) outs(%[[OUT_BUF4]] : tensor<256x1x1x16xf32>) permutation = [0, 2, 3, 1]
+// LINALG-CHECK:      %{{.+}} = hal.tensor.alias wait(%{{.+}}) => %[[OUT_T4]] : tensor<256x1x1x16xf32> to %[[ARG0]] : !hal.buffer_view
+//
+// AMDGPU-STATS-CHECK: "dispatch-count": 2
+// CPU-STATS-CHECK: "dispatch-count": 2
```
Comment (a-sidorova) [inline sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp]: After lowering we get two dispatch regions: convolution and sum reduction by batch. As far as I understand, this is expected behavior for  backward not unit-stride convolutions with weight gradient calculations.

Reference in "DecomposeComplexOps" pass in torch-mlir: [link](https://github.com/iree-org/torch-mlir/blob/b8f742b2aa0fafa065d37198f574c4d23aabd886/lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp#L5824)
Logs (mlir, command line, lowered mlir): [link](https://gist.github.com/a-sidorova/054db1fb0adacaf94e8ef8f1760d3e52)

File: sharkfuser/samples/convolution/conv_wgrad_nhwc_krsc_grouped_strided.cpp [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    // Compile
+    FUSILLI_REQUIRE_OK(graph->compile(handle, /*remove=*/true));
+
+    return std::make_tuple(graph, dyT, xT, dwT);
+  };
+
+  // Parameterize sample by backend and create device-specific handles.
+  std::shared_ptr<Handle> handlePtr;
+  // TODO(#2630): Add CPU backend when the SIGSEGV issue on CPU is fixed
```
Comment (sjain-stanford) [inline sharkfuser/samples/convolution/conv_wgrad_nhwc_krsc_grouped_strided.cpp]: nit: add the code and leave it commented: 
```
//  SECTION("cpu backend") {
//    handlePtr = std::make_shared<Handle>(
//        FUSILLI_REQUIRE_UNWRAP(Handle::create(Backend::CPU)));
//  }
```

File: sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp (hunk 1 above)
Comment (sjain-stanford) [reply sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp]: @IanWood1 please verify if this is expected. If not, let's add a TODO with the issue link.

File: sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+// TORCH-CHECK:       %permute_DW_val_2_conv_wgrad = torch.constant.int 3
+// TORCH-CHECK:       %permute_DW_val_3_conv_wgrad = torch.constant.int 1
+// TORCH-CHECK:       %permute_DW_conv_wgrad = torch.prim.ListConstruct %permute_DW_val_0_conv_wgrad, %permute_DW_val_1_conv_wgrad, %permute_DW_val_2_conv_wgrad, %permute_DW_val_3_conv_wgrad : (!torch.int, !torch.int, !torch.int, !torch.int) -> !torch.list<int>
+// TORCH-CHECK:       %result = torch.aten.permute %result_perm, %permute_DW_conv_wgrad : !torch.vtensor<[256,16,1,1],f32>, !torch.list<int> -> !torch.vtensor<[256,1,1,16],f32>
+// TORCH-CHECK:       torch.overwrite.tensor.contents %result overwrites %result_ : !torch.vtensor<[256,1,1,16],f32>, !torch.tensor<[256,1,1,16],f32>
+// TORCH-CHECK:       return
+// TORCH-CHECK:     }
+// TORCH-CHECK:   }
+//
+// LINALG-CHECK:    util.func public @main$async(%[[ARG0:.+]]: !hal.buffer_view, %[[ARG1:.+]]: !hal.buffer_view, %[[ARG2:.+]]: !hal.buffer_view, {{.+}}
```
[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp]: For this case though, I feel we can just drop the LINALG-CHECKS entirely, because:
1. The IR is sub-optimal and will go through significant changes with the torch-mlir revamp of backward conv lowering that you and others are working on). 
2. Tests on linalg transformations would normally live alongside the sources for those linalg transformations - so torch-mlir or IREE (we don't do any of that in fusilli as we stop at emitting torch dialect). 

We made an exception to 2 for some lit tests and added the linalg checks to verify certain assumptions we make in fusilli around layout transformations - like emitting transposes for NHWC -> NCHW don't incur an actual cost and are absorbed by the compiler simplifications in Torch->Linalg. Sorry for the confusion.

[the remark this answers is not in the document]
Comment (sjain-stanford) [reply sharkfuser/tests/lit/test_conv_wgrad_asm_emitter_nhwc_krsc_grouped_strided.cpp]: If you don't mind, I'll go ahead and land this as-is for now, and cleanup the LINALG checks in a follow-on.
